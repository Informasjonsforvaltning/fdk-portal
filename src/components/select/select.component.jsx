import React from 'react';
import PropTypes from 'prop-types';
import { map } from 'lodash';
import {
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  UncontrolledDropdown
} from 'reactstrap';

import localization from '../../lib/localization';

export class Select extends React.Component {
  constructor(props) {
    super(props);
    const { activeSort } = this.props;
    this.state = {
      selectedValue: activeSort
        ? activeSort.substring(0, activeSort.lastIndexOf('.'))
        : 'relevance'
    };
    this.onChange = this.onChange.bind(this);
  }

  onChange(e) {
    const { onChange } = this.props;
    onChange(e);
    this.setState({
      selectedValue: e.label
    });
  }

  render() {
    const { items } = this.props;
    const { selectedValue } = this.state;
    return (
      <UncontrolledDropdown
        id="search-result-dropdown-1"
        className="ml-auto"
        title={`${localization.sort.by} ${localization.sort[selectedValue]}`}
        aria-label={`Sorter sÃ¸keresultat, ${localization.sort.by} ${localization.sort[selectedValue]}`}
      >
        <DropdownToggle className="fdk-button" caret color="primary" title="">
          {localization.sort.by} {localization.sort[selectedValue]}
        </DropdownToggle>
        <DropdownMenu right className="fdk-dropdownmenu">
          {map(items, (item, idx) => {
            const { label } = item;
            const text = `${localization.sort.by} ${localization.sort[label]}`;
            return (
              <DropdownItem
                key={idx}
                onClick={() => this.onChange(item)}
                title=""
              >
                {text}
              </DropdownItem>
            );
          })}
        </DropdownMenu>
      </UncontrolledDropdown>
    );
  }
}

Select.defaultProps = {
  items: null,
  activeSort: null
};

Select.propTypes = {
  items: PropTypes.array,
  activeSort: PropTypes.string,
  onChange: PropTypes.func.isRequired
};
