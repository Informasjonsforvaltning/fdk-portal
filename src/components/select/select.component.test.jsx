import React from 'react';
import { shallow } from 'enzyme';
import { Select } from './select.component';

let onChange;
let items;

beforeEach(() => {
  onChange = jest.fn();
  items = [
    {
      label: 'relevance',
      field: '_score',
      order: 'asc',
      defaultOption: true
    },
    {
      label: 'title',
      field: 'title',
      order: 'asc'
    },
    {
      label: 'modified',
      field: 'modified',
      order: 'desc'
    },
    {
      label: 'publisher',
      field: 'publisher.name',
      order: 'asc'
    }
  ];
});

test('should render Select correctly with no active sort', () => {
  const wrapper = shallow(<Select onChange={onChange} items={items} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render Select correctly with active sort', () => {
  const wrapper = shallow(
    <Select onChange={onChange} items={items} activeSort="title.nb" />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should handle onChange', () => {
  const wrapper = shallow(<Select onChange={onChange} items={items} />);
  wrapper
    .find('DropdownItem')
    .first()
    .prop('onClick')(items[0]);
  expect(onChange).toHaveBeenLastCalledWith(items[0]);
});
