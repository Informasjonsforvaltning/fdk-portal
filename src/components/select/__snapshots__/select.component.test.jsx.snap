// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Select correctly with active sort 1`] = `
<UncontrolledDropdown
  aria-label="Sorter søkeresultat, Sort by title A-Z"
  className="ml-auto"
  id="search-result-dropdown-1"
  title="Sort by title A-Z"
>
  <DropdownToggle
    aria-haspopup={true}
    caret={true}
    className="fdk-button"
    color="primary"
    title=""
  >
    Sort by
     
    title A-Z
  </DropdownToggle>
  <DropdownMenu
    className="fdk-dropdownmenu"
    flip={true}
    right={true}
    tag="div"
  >
    <DropdownItem
      key="0"
      onClick={[Function]}
      tag="button"
      title=""
      toggle={true}
    >
      Sort by Relevance
    </DropdownItem>
    <DropdownItem
      key="1"
      onClick={[Function]}
      tag="button"
      title=""
      toggle={true}
    >
      Sort by title A-Z
    </DropdownItem>
    <DropdownItem
      key="2"
      onClick={[Function]}
      tag="button"
      title=""
      toggle={true}
    >
      Sort by Recently published
    </DropdownItem>
    <DropdownItem
      key="3"
      onClick={[Function]}
      tag="button"
      title=""
      toggle={true}
    >
      Sort by publisher A-Z
    </DropdownItem>
  </DropdownMenu>
</UncontrolledDropdown>
`;

exports[`should render Select correctly with no active sort 1`] = `
<UncontrolledDropdown
  aria-label="Sorter søkeresultat, Sort by Relevance"
  className="ml-auto"
  id="search-result-dropdown-1"
  title="Sort by Relevance"
>
  <DropdownToggle
    aria-haspopup={true}
    caret={true}
    className="fdk-button"
    color="primary"
    title=""
  >
    Sort by
     
    Relevance
  </DropdownToggle>
  <DropdownMenu
    className="fdk-dropdownmenu"
    flip={true}
    right={true}
    tag="div"
  >
    <DropdownItem
      key="0"
      onClick={[Function]}
      tag="button"
      title=""
      toggle={true}
    >
      Sort by Relevance
    </DropdownItem>
    <DropdownItem
      key="1"
      onClick={[Function]}
      tag="button"
      title=""
      toggle={true}
    >
      Sort by title A-Z
    </DropdownItem>
    <DropdownItem
      key="2"
      onClick={[Function]}
      tag="button"
      title=""
      toggle={true}
    >
      Sort by Recently published
    </DropdownItem>
    <DropdownItem
      key="3"
      onClick={[Function]}
      tag="button"
      title=""
      toggle={true}
    >
      Sort by publisher A-Z
    </DropdownItem>
  </DropdownMenu>
</UncontrolledDropdown>
`;
